File: src/engine/input-processor.ts
import type { CharResult, WordState } from "../types/stats.js";

export interface InputState {
  words: WordState[];
  currentWordIndex: number;
  cursorPosition: number;
  finished: boolean;
  historicalErrorChars: number;
  /** Total correct keypresses (never decremented by backspace) */
  keypressCorrect: number;
  /** Total incorrect keypresses (never decremented by backspace) */
  keypressIncorrect: number;
}

export function createInputState(words: string[]): InputState {
  return {
    words: words.map((word) => ({
      word,
      typed: "",
      chars: [],
      completed: false,
    })),
    currentWordIndex: 0,
    cursorPosition: 0,
    finished: false,
    historicalErrorChars: 0,
    keypressCorrect: 0,
    keypressIncorrect: 0,
  };
}

export function processChar(state: InputState, char: string): InputState {
  if (state.finished) return state;

  const words = [...state.words];
  const idx = state.currentWordIndex;
  const current = { ...words[idx] };
  const pos = current.typed.length;
  const expected = current.word[pos] ?? "";
  const isExtra = pos >= current.word.length;

  const result: CharResult = {
    char,
    expected,
    correct: !isExtra && char === expected,
    extra: isExtra,
    timestamp: performance.now(),
  };

  current.typed = current.typed + char;
  current.chars = [...current.chars, result];
  words[idx] = current;

  const isCorrectKeypress = result.correct;

  return {
    ...state,
    words,
    cursorPosition: current.typed.length,
    historicalErrorChars:
      state.historicalErrorChars + (isCorrectKeypress ? 0 : 1),
    keypressCorrect: state.keypressCorrect + (isCorrectKeypress ? 1 : 0),
    keypressIncorrect: state.keypressIncorrect + (isCorrectKeypress ? 0 : 1),
  };
}

export function processSpace(state: InputState): InputState {
  if (state.finished) return state;

  const words = [...state.words];
  const idx = state.currentWordIndex;

  // Don't advance if nothing typed in current word
  if (words[idx].typed.length === 0) return state;

  const current = { ...words[idx] };
  current.completed = true;
  words[idx] = current;

  const nextIndex = idx + 1;
  const finished = nextIndex >= words.length;

  // Space is a correct keypress only if the word was typed correctly
  const spaceCorrect = current.typed === current.word;

  return {
    ...state,
    words,
    currentWordIndex: finished ? idx : nextIndex,
    cursorPosition: 0,
    finished,
    keypressCorrect: state.keypressCorrect + (spaceCorrect ? 1 : 0),
    keypressIncorrect: state.keypressIncorrect + (spaceCorrect ? 0 : 1),
  };
}

export function processBackspace(state: InputState): InputState {
  if (state.finished) return state;

  const words = [...state.words];
  const idx = state.currentWordIndex;
  const current = { ...words[idx] };

  if (current.typed.length > 0) {
    // Delete last character in current word
    current.typed = current.typed.slice(0, -1);
    current.chars = current.chars.slice(0, -1);
    words[idx] = current;

    return {
      ...state,
      words,
      cursorPosition: current.typed.length,
    };
  }

  // If at start of word and not the first word, go back to previous word
  if (idx > 0) {
    const prev = { ...words[idx - 1] };
    // Only allow going back to a word that was just completed
    if (prev.completed) {
      prev.completed = false;
      words[idx - 1] = prev;

      return {
        ...state,
        words,
        currentWordIndex: idx - 1,
        cursorPosition: prev.typed.length,
      };
    }
  }

  return state;
}

export function processCtrlBackspace(state: InputState): InputState {
  if (state.finished) return state;

  const words = [...state.words];
  const idx = state.currentWordIndex;
  const current = { ...words[idx] };

  current.typed = "";
  current.chars = [];
  words[idx] = current;

  return {
    ...state,
    words,
    cursorPosition: 0,
  };
}

export function processKeystroke(
  state: InputState,
  input: string,
  key: { backspace?: boolean; ctrl?: boolean },
): InputState {
  if (key.backspace) {
    if (key.ctrl) {
      return processCtrlBackspace(state);
    }
    return processBackspace(state);
  }

  if (input === " ") {
    return processSpace(state);
  }

  // Only process printable characters (single chars)
  if (input.length === 1 && input >= " ") {
    return processChar(state, input);
  }

  return state;
}

File: src/engine/input_processor.rs
use crate::types::{CharResult, InputState, WordState};
use std::time::Instant;

pub fn create_input_state(words: &[String]) -> InputState {
    InputState {
        words: words
            .iter()
            .map(|w| WordState {
                word: w.clone(),
                typed: String::new(),
                chars: Vec::new(),
                completed: false,
            })
            .collect(),
        current_word_index: 0,
        cursor_position: 0,
        finished: false,
        historical_error_chars: 0,
        keypress_correct: 0,
        keypress_incorrect: 0,
    }
}

pub fn process_char(mut state: InputState, ch: char) -> InputState {
    if state.finished {
        return state;
    }

    let idx = state.current_word_index;
    let current = &mut state.words[idx];
    let pos = current.typed.len();
    let expected = current.word.chars().nth(pos).unwrap_or('\0');
    let is_extra = pos >= current.word.len();

    let correct = !is_extra && ch == expected;

    let result = CharResult {
        char: ch,
        expected,
        correct,
        extra: is_extra,
        timestamp: Instant::now(),
    };

    current.typed.push(ch);
    current.chars.push(result);

    state.cursor_position = current.typed.len();
    if !correct {
        state.historical_error_chars += 1;
    }
    if correct {
        state.keypress_correct += 1;
    } else {
        state.keypress_incorrect += 1;
    }

    state
}

pub fn process_space(mut state: InputState) -> InputState {
    if state.finished {
        return state;
    }

    let idx = state.current_word_index;

    // Don't advance if nothing typed in current word
    if state.words[idx].typed.is_empty() {
        return state;
    }

    let space_correct = state.words[idx].typed == state.words[idx].word;
    state.words[idx].completed = true;

    let next_index = idx + 1;
    let finished = next_index >= state.words.len();

    state.current_word_index = if finished { idx } else { next_index };
    state.cursor_position = 0;
    state.finished = finished;

    if space_correct {
        state.keypress_correct += 1;
    } else {
        state.keypress_incorrect += 1;
    }

    state
}

pub fn process_backspace(mut state: InputState) -> InputState {
    if state.finished {
        return state;
    }

    let idx = state.current_word_index;

    if !state.words[idx].typed.is_empty() {
        // Delete last character in current word
        state.words[idx].typed.pop();
        state.words[idx].chars.pop();
        state.cursor_position = state.words[idx].typed.len();
        return state;
    }

    // If at start of word and not the first word, go back to previous word
    if idx > 0 && state.words[idx - 1].completed {
        state.words[idx - 1].completed = false;
        state.current_word_index = idx - 1;
        state.cursor_position = state.words[idx - 1].typed.len();
    }

    state
}

pub fn process_ctrl_backspace(mut state: InputState) -> InputState {
    if state.finished {
        return state;
    }

    let idx = state.current_word_index;
    state.words[idx].typed.clear();
    state.words[idx].chars.clear();
    state.cursor_position = 0;

    state
}

pub fn process_keystroke(state: InputState, input: &str, backspace: bool, ctrl: bool) -> InputState {
    if backspace {
        if ctrl {
            return process_ctrl_backspace(state);
        }
        return process_backspace(state);
    }

    if input == " " {
        return process_space(state);
    }

    // Only process printable characters (single chars)
    let chars: Vec<char> = input.chars().collect();
    if chars.len() == 1 && chars[0] >= ' ' {
        return process_char(state, chars[0]);
    }

    state
}
